students/jeff_shabani/session07/html_render.py

```diff
@@ -7,22 +7,145 @@
 
 # This is the framework for the base class
 class Element(object):
+    indent = ' ' * 2
+    tag = 'html'
 
-    def __init__(self, content=None):
-        self.content = str(content)
+    def __init__(self, content=None, **attrs):
+        self.attrs = attrs
+        if content:
+            self.content = [content]
+        else:
+            self.content = []
+
+    def _front_tag(self):
+        """Creates the front/opening tag"""
+        return f'<{self.tag}>'
+
+    def _end_tag(self):
+        """Creates the ending tag"""
+        return f'</{self.tag}>'
 
     def append(self, new_content):
-        result = list()
-        result.append(new_content)
-        for i in result:
-            self.content +=i
-        return self.content
+        self.content.append(new_content)
+
+    def render(self, file_name, cur_indent=''):
+        # Writes the opening tag
+        file_name.write(f'{self._front_tag()}\n')
+        for content_line in self.content:
+            file_name.write(f'{content_line}\n')
```

Comment:
You need to consider that not all your content elements will be strings, some will be class instances, so instead writing to file_name, you will need to call the instance's own _render_ method (with a different indentation, of course). You can use _hasattr()_ to see if _content_line_ has a _render()_ attribute. If that's the case, call _content_line.render(...)_.

Reply:
Fixed. This seems so simple in hind-sight. I really causes me concern that I didn't pick up on this myself, though.